bigframes/core/blocks.py

```diff
@@ -2011,6 +2012,61 @@ def _is_monotonic(
         self._stats_cache[column_name].update({op_name: result})
         return result
 
+    def _get_rows_as_json_values(self) -> Block:
+        sql, index_column_ids, _ = self.to_sql_query(include_index=True)
+
+        # all column names, types and values
+        all_column_names = index_column_ids + [col for col in self.column_labels]
+        column_names_csv = ", ".join([f'"{col}"' for col in all_column_names])
+        column_references_csv = ", ".join(
+            [f"CAST({col} AS STRING)" for col in all_column_names]
+        )
+        all_column_types = list(self.index.dtypes) + list(self.dtypes)
+        column_types_csv = ", ".join([f'"{col}"' for col in all_column_types])
+
+        # index column names
+        index_column_names_csv = ", ".join([f'"{col}"' for col in index_column_ids])
+
+        row_json_column_name = guid.generate_guid()
+        select_columns = index_column_ids + [row_json_column_name]
+        select_columns_csv = ", ".join(select_columns)
+        json_sql = f"""\
+With T0 AS (
+{textwrap.indent(sql, "    ")}
+),
+T1 AS (
+    SELECT *,
+           JSON_OBJECT(
+               "names", [{column_names_csv}],
+               "types", [{column_types_csv}],
+               "values", [{column_references_csv}],
+               "index", [{index_column_names_csv}]
+           ) AS {row_json_column_name} FROM T0
+)
+SELECT {select_columns_csv} FROM T1
+"""
+        ibis_table = self.session.ibis_client.sql(json_sql)
+        order_for_ibis_table = ordering.ExpressionOrdering(
+            ordering_value_columns=tuple(
+                [ordering.ascending_over(column_id) for column_id in index_column_ids]
+            ),
+            total_ordering_columns=frozenset(index_column_ids),
+        )
+        expr = core.ArrayValue.from_ibis(
+            self.session,
+            ibis_table,
+            [ibis_table[col] for col in select_columns],
+            hidden_ordering_columns=[],
+            ordering=order_for_ibis_table,
+        )
```

Comment:
by creating a brand new expression object, we decouple it from the original data sources, make some optimizations more difficult, most notably an inability to use the implicit joiner. Lets try to get this working as a scalar operation

Reply:
Left a TODO for doing it in another change, tracked by internal issue 338218579.